sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala

```diff
@@ -1753,15 +1753,26 @@ class DataFrame private[sql](
    * Converts a JavaRDD to a PythonRDD.
    */
   protected[sql] def javaToPython: JavaRDD[Array[Byte]] = {
-    val structType = schema  // capture it for closure
-    val rdd = queryExecution.toRdd.map(EvaluatePython.toJava(_, structType))
-    EvaluatePython.javaToPython(rdd)
+    if (isOutputPickled) {
+      queryExecution.toRdd.map(_.getBinary(0))
```

Comment:
For `groupByKey`, we need the JVM to do grouping for us, i.e. cluster by key and sort by key. If the grouped Dataset contains custom objects, we need to send it JVM and keep it as binary while do the grouping. As we need to append the key data to every input row, we should use un-batched pickler here.